app/adapters/abstract_adapter.rb

```diff
@@ -0,0 +1,294 @@
+# frozen_string_literal: true
+
+require 'uri'
+require 'httpclient'
+require 'mutex_m'
+
+# KeycloakAdapter adapter to create/update/delete Clients on using the KeycloakAdapter Client Registration API.
+class AbstractAdapter
+  def self.build_client(*)
+    raise NotImplementedError, __method__
+  end
+
+  attr_reader :endpoint
+
+  def initialize(endpoint, authentication: nil)
+    endpoint = EndpointConfiguration.new(endpoint)
+    @http_client = build_http_client(endpoint)
+    @oidc = OIDC.new(endpoint, http_client)
+    @oidc.access_token = authentication if authentication
+    @endpoint = endpoint.issuer
+  end
+
+  def authentication=(value)
+    oidc.access_token = value
+  end
+
+  def authentication
+    oidc.access_token.token
+  end
+
+  def create_client(_)
+    raise NotImplementedError, __method__
+  end
+
+  def read_client(_)
+    raise NotImplementedError, __method__
+  end
+
+  def update_client(_)
+    raise NotImplementedError, __method__
+  end
+
+  def delete_client(_)
+    raise NotImplementedError, __method__
+  end
+
+  def test
+    raise NotImplementedError, __method__
+  end
+
+  protected
+
+  attr_reader :oidc
+
+  def headers
+    oidc.headers
+  end
+
+  JSON_TYPE = Mime[:json]
+  private_constant :JSON_TYPE
+
+  NULL_TYPE = Mime::Type.lookup(nil)
+
+  attr_reader :http_client
+
+  def build_http_client(endpoint)
+    HTTPClient.new do
+      self.debug_dev = $stderr if ENV.fetch('DEBUG', '0') == '1'
+
+      self.set_auth endpoint, *endpoint.auth
+
+      Rails.application.config.x.http_client.deep_symbolize_keys
+          .slice(:connect_timeout, :send_timeout, :receive_timeout).each do |key, value|
+        self.public_send("#{key}=", value)
+      end
+    end
+  end
+
+  def parse(response)
+    body = self.class.parse_response(response)
+
+    raise InvalidResponseError, { response: response, message: body } unless response.ok?
+
+    params = body.try(:to_h) or return # no need to create client if there are no attributes
```

Comment:
If the content-type is NULL_TYPE, `body` will be a `String`
https://github.com/3scale/zync/pull/196/files#diff-cbd4720b5437174ca7b9bcc0a46568adR99

Reply:
Which is fine, because it will return `nil` and return.